File: kernel/i386/paging.c
#include <arch/i386.h>
#include <pirix/kprint.h>
#include <pirix/paging.h>
#include <pirix/memory.h>
#include <pirix/irq.h>
#include <pirix/kernel.h>
#include <pirix/string.h>

unsigned* page_dir = (unsigned*)0xffbff000;
unsigned* page_tables = (unsigned*)0xffc00000;

static paging_context current_context = NULL;

registers* paging_fault(registers* regs) {
    unsigned long addr;
    asm volatile("mov %%cr2, %0" : "=r"(addr));
    kprintf("page fault at %p, eip: %p\n", addr, regs->eip);
    panic(0);
    return regs;
}

void paging_init() {
    irq_register(14, &paging_fault);

    /*
     * initialize kernel context
     *
     * addresses 0x0 - 0xc0000000 are currently identity mapped,
     * memory_alloc should return a free memory address from this
     * block, otherwise we are in serious trouble.
     */

    unsigned* kcontext = (unsigned*)memory_alloc();

    memset(kcontext, 0, 768*4);

    // create page tables from 0xc0000000
    for (int i = 0; i < 255; i++) {
        unsigned* table = (unsigned*)memory_alloc();
        kcontext[i+768] = (unsigned long)table | 0x3;

        memset(table, 0, 0x1000);

        if (i == 0) {
            // map 0xc0000000 to the first 4MB
            for (int j = 0; j < 1024; j++) {
                table[j] = j*0x1000 | 0x103;
            }
        }
        else if (i == 254) {
            table[1023] = (unsigned long)kcontext | 0x3;
        }
    }

    kcontext[1023] = (unsigned long)kcontext | 0x3;

    paging_activate_context(kcontext);
}

paging_context paging_create_context() {
    unsigned* pcontext = (unsigned*)memory_alloc();
    unsigned* context = paging_map_kernel((unsigned long)pcontext);

    unsigned* ptable = (unsigned*)memory_alloc();
    unsigned* table = paging_map_kernel((unsigned long)ptable);

    memset(table, 0, 0x1000);
    memset(context, 0, 0xc00);
    memcpy(context+768, page_dir+768, 0x3f8);

    table[1023] = (unsigned long)pcontext | 0x3;
    context[1022] = (unsigned long)ptable | 0x3;
    context[1023] = (unsigned long)pcontext | 0x3;

    paging_unmap_kernel((unsigned long)table);
    paging_unmap_kernel((unsigned long)context);

    return pcontext;
}

int paging_map(paging_context context, unsigned long virt, unsigned long phys, unsigned access) {
    unsigned* dir, *table;
    unsigned long pdidx = virt >> 22;
    unsigned long ptidx = (virt >> 12) & 0x3ff;

    access = (access == PAGE_PERM_USER) ? 0x4 : 0x0;

    if (!context) context = current_context;

    if (context == current_context && page_dir[pdidx]) {
        page_tables[ptidx+0x400*pdidx] = phys | access | 0x3;
        invlpg(virt);
        return 0;
    }

    dir = (unsigned*)paging_map_kernel((unsigned long)context);

    // create page table if unexistent
    if (!(dir[pdidx] & 0x1)) {
        unsigned* ptable = (unsigned*)memory_alloc();
        table = paging_map_kernel((unsigned long)ptable);
        memset(table, 0, 0x1000);
        dir[pdidx] = (unsigned long)ptable | access | 0x3;
    }
    else {
        table = paging_map_kernel(dir[pdidx] & 0xfffff000);
    }

    table[ptidx] = phys | access | 0x3;
    paging_unmap_kernel((unsigned long)table);
    paging_unmap_kernel((unsigned long)dir);

    invlpg(virt);

    return 0;
}

void* paging_map_kernel(unsigned long phys) {
    for (unsigned long virt = 0xd0000000; virt < 0xe0000000; virt += 0x1000) {
        unsigned* entry = &page_tables[virt >> 12];

        if (!(*entry & 0x1)) {
            *entry = phys | 0x103;
            invlpg(virt);
            return (void*)(virt + (phys & 0xfff));
        }
    }

    panic("kernel mapping area full");
    return 0;
}

void paging_unmap_kernel(unsigned long virt) {
    page_tables[virt >> 12] = 0;
    invlpg(virt);
}

void paging_activate_context(paging_context context) {
    if (context == current_context) return;
    asm volatile("mov %0, %%cr3" :: "r"(context) : "memory");
    current_context = context;
}

void* paging_getphys(unsigned long virt) {
    unsigned long pdidx = virt >> 22;
    unsigned long ptidx = (virt >> 12) & 0x3FF + 0x400*pdidx;

    if (!(page_dir[pdidx] & 0x1)) return 0;
    if (!(page_tables[ptidx] & 0x1)) return 0;
    return (void*)((page_tables[ptidx] & ~0xFFF) + (virt & 0xFFF));
}

File: arch/i386/paging.rs
use irq;
use arch::cpu;
use mem::frame;
use core::mem::size_of;
use core::ptr::write_bytes;

#[repr(packed)] struct PageTable { entries: [usize; 1024] }
#[repr(packed)] struct PageDir { entries: [usize; 1024] }

pub struct PagingContext(*mut PageDir);

extern {
    static mut boot_page_dir: PageDir;
    static mut kernel_page_tables: [PageTable; 256];
}

struct MappedTables {
    page_dir: *mut PageDir,
    page_tables: *mut PageTable
}

unsafe impl Sync for MappedTables {}

static mapped: MappedTables = MappedTables {
    page_dir: 0xfffff000 as *mut PageDir,
    page_tables: 0xffc00000 as *mut PageTable
};

impl PageDir {
    pub unsafe fn new() -> *mut PageDir {
        let dir: *mut PageDir = frame::alloc();
        let vdir = kernel_map(dir);

        let table: *mut PageTable = frame::alloc();
        let vtable = kernel_map(table);

        (*vtable).clear();
        (*vdir).clear();

        // copy kernel pages
        for i in 768..1024 {
            let entry = (*mapped.page_dir).get(i);
            (*vdir).set(i, entry.0, entry.1);
        }

        (*vtable).set(1023, dir as usize, 0x3);
        (*vdir).set(1022, table, 0x3);
        (*vdir).set(1023, dir as *mut PageTable, 0x3);

        kernel_unmap(vtable);
        kernel_unmap(vdir);

        return dir;
    }

    pub unsafe fn mapped() -> *mut PageDir {
        mapped.page_dir as *mut PageDir
    }

    pub unsafe fn clear(&mut self) {
        write_bytes(self.entries.as_mut_ptr(), 0, 1024);
    }

    pub fn set(&mut self, index: usize, table: *mut PageTable, flags: usize) {
        self.entries[index] = table as usize | flags;
    }

    pub fn get(&self, index: usize) -> (*mut PageTable, usize) {
        ((self.entries[index] as usize & 0xfffff000) as *mut PageTable,
         self.entries[index] as usize & 0xfff)
    }
}

impl PageTable {
    pub fn new() -> *mut PageTable {
        return frame::alloc();
    }

    pub unsafe fn at(index: usize) -> *mut PageTable {
        mapped.page_tables.offset(index as isize) as *mut PageTable
    }

    pub unsafe fn clear(&mut self) {
        write_bytes(self.entries.as_mut_ptr(), 0, 1024);
    }

    pub fn set(&mut self, index: usize, page: usize, flags: usize) {
        self.entries[index] = page | flags;
    }

    pub fn get(&self, index: usize) -> usize {
        self.entries[index]
    }
}

fn page_fault(state: &mut cpu::State) {
    let addr: u32;
    unsafe { asm!("mov %cr2, $0" : "=r"(addr)); }

    log!("page fault at 0x{:x}:", addr);
    if state.err & 1 == 1 { log!(" - protection violation"); }
    else { log!(" - page does not exist"); }

    if state.err & 2 == 2 { log!(" - error writing"); }
    else { log!(" - error reading"); }

    if state.err & 4 == 4 { log!(" - occured in user mode"); }
    else { log!(" - occured in supervisor mode"); }

    if state.err & 16 == 16 { log!(" - caused by instruction fetch"); }

    panic!("unhandled page fault!");
}

pub unsafe fn init() {
    irq::register(0x0E, page_fault);

    let physdir: *mut PageDir = &mut boot_page_dir;
    let dir = ((physdir as usize) + 0xc0000000) as *mut PageDir;

    // create kernel pages
    for i in 768..1023 {
        let table = kernel_page_tables.get_mut(i-768).unwrap() as *mut PageTable;

        (*table).clear();

        if i == 768 {
            for j in 0..1024 {
                (*table).set(j, j*0x1000, 0x103);
            }
        }
        else if i == 1022 {
            (*table).set(1023, dir as usize, 0x3);
        }

        let phystable = (table as usize - 0xc0000000) as *mut PageTable;
        (*dir).set(i, phystable, 0x3);
    }

    (*dir).set(1023, physdir as *mut PageTable, 0x3);

    activate_pagedir(physdir);
}

pub unsafe fn map(virt: usize, phys: usize, flags: usize) {
    let dir = PageDir::mapped();
    let pdidx = virt >> 22;
    let ptidx = (virt >> 12) & 0x3ff;

    if (*dir).get(pdidx).1 & 0x1 == 0 {
        let table = PageTable::new();
        (*dir).set(pdidx, table, flags | 0x3);
        let vtable = PageTable::at(pdidx);
        invlpg(vtable as usize);
        (*vtable).clear();
    }

    let table = PageTable::at(pdidx);
    (*table).set(ptidx, phys, flags | 0x3);

    invlpg(virt);
}

pub unsafe fn unmap(virt: usize) {
    let pdidx = virt >> 22;
    let ptidx = (virt >> 12) & 0x3ff;

    let table = PageTable::at(pdidx);
    (*table).set(ptidx, 0, 0);

    invlpg(virt);
}

pub unsafe fn getphys<T>(virt: *mut T) -> *mut T {
    let virt = virt as usize;

    let pdidx = virt >> 22;
    let ptidx = (virt >> 12) & 0x3ff;

    let table = PageTable::at(pdidx);
    let entry = (*table).get(ptidx);

    return ((entry & 0xfffff000) | (virt & 0xfff)) as *mut T;
}

fn calculate_page_count<T>(addr: *mut T) -> usize {
    let size = size_of::<T>();
    let mut page_count = (size+0x1000-1) / 0x1000;

    if (addr as usize % 0x1000) + size > 0x1000 {
        page_count += 1;
    }

    return page_count;
}

unsafe fn find_mapping_area(page_count: usize) -> usize {
    let mut continous_start = 0;
    let mut continous_count = 0;

    for virt in (0xd0000000..0xe0000000).step_by(0x1000) {
        let table = PageTable::at(virt >> 22);
        let index = (virt >> 12) & 0x3ff;

        if (*table).get(index) & 0x1 == 0x1 {
            continous_start = 0;
            continue;
        }

        if continous_start == 0 {
            continous_start = virt;
            continous_count = 1;
        }
        else {
            continous_count += 1;
        }

        if continous_count >= page_count {
            return continous_start;
        }
    }

    return 0;
}

pub unsafe fn kernel_map<T>(addr: *mut T) -> *mut T {
    let page_count = calculate_page_count(addr);
    let mapping_area = find_mapping_area(page_count);

    if mapping_area == 0 {
        panic!("kernel mapping area full");
    }

    for i in 0..page_count {
        let virt = mapping_area + i*0x1000;
        map(virt, addr as usize + i*0x1000, 0x103);
    }

    let vaddr = mapping_area + (addr as usize & 0xfff);

    return vaddr as *mut T;
}

pub unsafe fn kernel_unmap<T>(addr: *mut T) {
    let page_count = calculate_page_count(addr);

    for i in 0..page_count {
        let virt = (addr as usize) + i*0x1000;
        unmap(virt);
    }
}

impl PagingContext {
    pub fn new() -> PagingContext {
        PagingContext(unsafe { PageDir::new() })
    }

    pub fn activate(&self) {
        match *self {
            PagingContext(dir) => unsafe { activate_pagedir(dir) }
        }
    }
}

unsafe fn activate_pagedir(dir: *mut PageDir) {
    asm!("mov $0, %cr3" :: "r"(dir));
}

unsafe fn invlpg(addr: usize) {
    asm!("invlpg [$0]" :: "r"(addr) : "memory" : "volatile", "intel");
}
